src/main/java/pl/allegro/tech/opel/DeclarationExpressionNode.java

```diff
@@ -0,0 +1,26 @@
+package pl.allegro.tech.opel;
+
+import java.util.concurrent.CompletableFuture;
+
+public class DeclarationExpressionNode implements ExpressionNode {
+    private final IdentifierExpressionNode identifier;
+    private final ExpressionNode expression;
+
+    public DeclarationExpressionNode(ExpressionNode identifier, ExpressionNode expression) {
+        this.identifier = (IdentifierExpressionNode) identifier;
+        this.expression = expression;
+    }
+
+    public IdentifierExpressionNode getIdentifier() {
+        return identifier;
+    }
+
+    public ExpressionNode getExpression() {
+        return expression;
+    }
+
+    @Override
+    public CompletableFuture<?> getValue(EvalContext context) {
```

Comment:
in grammar you add Declarations beside Expression so DeclarationExpressionNode should not implements ExpressionNode (Declaration is not an Expression) and problem with this UnsupportedOperationException disappear :)